ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -717,6 +728,74 @@ impl fmt::Display for AddInstruction {
     }
 }
 
+/// The operand for a [Instruction::Icmp]
+///
+/// # Semantics
+///
+/// Compares two operands.
+///
+#[derive(Debug)]
+pub struct IcmpInstruction {
+    op1: PackedOperand,
+    op2: PackedOperand,
+}
+
+impl IcmpInstruction {
+    pub(crate) fn new(op1: Operand, op2: Operand) -> Self {
+        Self {
+            op1: PackedOperand::new(&op1),
+            op2: PackedOperand::new(&op2),
+        }
+    }
+
+    fn op1(&self) -> Operand {
+        self.op1.get()
+    }
+
+    fn op2(&self) -> Operand {
+        self.op2.get()
+    }
+}
+
+impl fmt::Display for IcmpInstruction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Icmp {}, {}", self.op1(), self.op2())
+    }
+}
+
+/// The operand for a [Instruction::Guard]
+///
+/// # Semantics
+///
+/// Guards a trace against diverging execution.
+///
+#[derive(Debug)]
+pub struct GuardInstruction {
+    /// The condition to guard against.
+    cond: PackedOperand,
+    /// The expected outcome of the condition.
+    expect: bool,
+}
+
+impl GuardInstruction {
+    pub(crate) fn new(cond: Operand, expect: bool) -> Self {
+        GuardInstruction {
```

Comment:
debug assert that the condition is a bool?

Reply:
Does LLVM guarantee that? `BranchInst.getCondition()` returns `Value`.

Reply:
It's signature is `Value *getCondition()`, so I guess so.

Reply:
Well a "bool" is an `i1` in LLVM IR (and in ours too, right)?

Maybe I meant to say: check the condition is a 1-bit integer?

Reply:
Yes, but I what I meant is that I haven't seen LLVM enforcing this. Since `BranchInst` just takes a `Value` that could also be a `i8, i32, i64` etc.

Reply:
Ah never mind. It's defined in the langref: https://llvm.org/docs/LangRef.html#br-instruction

Reply:
Unfortunately, we don't store the type of the condition (why waste precious bytes when we know icmp returns `i1`) and we know that the AOT module must be verified so we know that condition must be `i1`.

Reply:
There are two LLVM instructions that might influence our decision:

 - icmp
 - br

They both deal with `i1`:

# [conditional `br`](https://llvm.org/docs/LangRef.html#br-instruction)

```
br i1 <cond>, label <iftrue>, label <iffalse>
```

> The conditional branch form of the ‘br’ instruction takes a single ‘i1’

# [`icmp`](https://llvm.org/docs/LangRef.html#icmp-instruction)

```
<result> = icmp <cond> <ty> <op1>, <op2>   ; yields i1 or <N x i1>:result
```

It probably takes a generic `Value` so that you can branch on locals or globals etc.

(We are talking about the *type* of a Value)

Reply:
I understand, but a) the codegen tests construct IR that didn't come from LLVM, and b) aren't assertions for testing our assumptions after all?

I won't die on this hill, but seems worthwhile to me. It's up to you.

Reply:
I started to implement this and realised that a) it's a bit of a pain to do this in the trace builder and b) this isn't the right place to do this. Since we can give the codegen a hand written trace, it should be the codegen to assert this, which would also cover mistakes hand written AOT IR. Though at some point it might be useful to have an AOT IR verifier as well.

Reply:
OK